### ISSUE TITLE
[Python3 target] BufferedTokenStream. get(self, index:int) does not check valid range of index

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#4427]
This is the code for the function for Python3. https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/Python3/src/antlr4/BufferedTokenStream.py#L80-L81

This code does not check whether `index` is in range.

For comparison, the other targets perform the check.

https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/CSharp/src/BufferedTokenStream.cs#L257-L261

https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/Java/src/org/antlr/v4/runtime/BufferedTokenStream.java#L185-L188

Regardless of the fact that no code should call `get()` with an invalid index, the Python3 code semantics should be identical across targets.

I have not checked all targets, but that should be done when a PR is submitted.

**Comment by ericvergnaud** [comment]
I don’t disagree with the principle but in this particular case the Java code is redundant since ArrayList.get already performs a bounds check, so the code slows down the runtime for the sole purpose of customizing the exception message, when this should never happen in the first placeI’d rather remove the redundant bounds check in Java and C#Envoyé de mon iPhoneLe 26 sept. 2023 à 02:29, Ken Domino ***@***.***> a écrit :﻿
This is the code for the function for Python3. https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/Python3/src/antlr4/BufferedTokenStream.py#L80-L81
This code does not check whether index is in range.
For comparison, the other targets perform the check.
https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/CSharp/src/BufferedTokenStream.cs#L257-L261
https://github.com/antlr/antlr4/blob/0eb38a02d7aa67147bdba23287501d4e5540dfc1/runtime/Java/src/org/antlr/v4/runtime/BufferedTokenStream.java#L185-L188
Regardless of the fact that no code should call get() with an invalid index, the Python3 code semantics should be identical across targets.
I have not checked all targets, but that should be done when a PR is submitted.

**Comment by kaby76** [comment]
Good point. But Python3 has the `-O` option. I will gather statistics to test the hypothesis.
